2/contrib/jenkins/install-plugins.sh

```diff
@@ -233,23 +235,45 @@ function resolveDependencies() {
     for d in "${array[@]}"
     do
         plugin="$(cut -d':' -f1 - <<< "$d")"
+	#
+	# Note, matrix-auth plugin notes cloudbees-folder as optional in the archive, but then failed to
+	# load, citing a dependency that is too old, during testing ... so we will download optional dependencies
+	#
+	local versionFromPluginParam
         if [[ $d == *"resolution:=optional"* ]]; then
-            echo "Skipping optional dependency $plugin"
-        else
-            local pluginInstalled
-            if pluginInstalled="$(echo "${bundledPlugins}" | grep "^${plugin}:")"; then
-                pluginInstalled="${pluginInstalled//[$'\r']}"
-                local versionInstalled; versionInstalled=$(versionFromPlugin "${pluginInstalled}")
-                local minVersion; minVersion=$(versionFromPlugin "${d}")
-                if versionLT "${versionInstalled}" "${minVersion}"; then
-                    echo "Upgrading bundled dependency $d ($minVersion > $versionInstalled)"
-                    download "$plugin" &
-                else
-                    echo "Skipping already bundled dependency $d ($minVersion <= $versionInstalled)"
-                fi
+            #echo "Skipping optional dependency $plugin"
+	    versionFromPluginParam="$(cut -d';' -f1 - <<< "$d")"
+	else
+	    versionFromPluginParam=$d
+	fi
+        local pluginInstalled
+        local minVersion; minVersion=$(versionFromPlugin "${versionFromPluginParam}")
+
+	set +o pipefail
+	local filename; filename=$(getArchiveFilename "$plugin")
+	local previouslyDownloadedVersion; previouslyDownloadedVersion=$(get_plugin_version $filename)
+	set -o pipefail
+	    
+        if pluginInstalled="$(echo "${bundledPlugins}" | grep "^${plugin}:")"; then
+            pluginInstalled="${pluginInstalled//[$'\r']}"
+            local versionInstalled; versionInstalled=$(versionFromPlugin "${pluginInstalled}")
+            if versionLT "${versionInstalled}" "${minVersion}"; then
+                echo "Upgrading bundled dependency $d ($minVersion > $versionInstalled)"
+                download "$plugin" "$minVersion" "true"
             else
-                download "$plugin" &
+                echo "Skipping already bundled dependency $d ($minVersion <= $versionInstalled)"
             fi
+	    continue
+        fi
+
+	if [[ ! $previouslyDownloadedVersion ]]; then
+	    echo "Downloading previously uninstalled dependency plugin $plugin version $minVersion"
```

Comment:
i'm still struggling w/ this message/logic.  how/why is this different from a plugin dependency that was never installed?  

Some comments around this and the plug above that computes the previouslyDownloadedVersion value would help me.